---
title: "My report  for analysis X (project title)"

author: Nina Dombrowski
date: 12/18/2023

format:
  html:
    embed-resources: true

toc: true
toc-depth: 2

execute:
  eval: false

engine: knitr
---

## Description

The goal of this workflow is to analyse sequencing data received on 23.01.2022 from Data Science Sequencing Centre. The sequencing data consists of 8 samples, 4 control samples and 4 sulfur-treatments. The code below documents the cleaning of the sequences and the analyses run in QIIME.

## Pre-requisites

All analyses were run on an HPC with an CentOS 7 operating system.

Dependencies:

-   FastQC v0.11.9 (pre-installed)
-   seqkit v2.6.1
-   ...

Installation of software that was not pre-installed:

```{bash}
mamba create -n seqkit_2.6.1
mamba install -n seqkit_2.6.1 -c bioconda seqkit=2.6.1
```

## Analysis workflow

### Setup working directory

```{bash}
#path to working directory
cd /path_to_folder_you_analyse_the_data

#activate conda environment 
mamba activate seqkit_2.6.1

#custom variables 
download_link="https://github.com/ndombrowski/cli_workshop/raw/main/data/seq_project.tar.gz"
```

### Download the sequencing data

```{bash}
mkdir data
mkdir scripts 
mkdir logs 

#download data using link provided by the sequencing center on 23.01.2023
wget -P data $download_link
tar -xvf data/seq_project.tar.gz -C data
rm data/*.tar.gz

#generate a list for the files we work with 
ls data/seq_project/*/*.gz | cut -f4 -d "/" > samples.txt
```

### Perform sanity checks

```{bash}
#we work with 8 files
wc -l samples.txt

#we have as many R1 as we have R2 files
ls data/seq_project/*/*R1* | wc -l
ls data/seq_project/*/*R2* | wc -l

#the sequence header looks as expected
zcat data/seq_project/*/*gz | head

#R1 and R2 reads have the same number of reads 
for i in `cat samples.txt`; do 
    echo "$i: $(( $(zcat data/seq_project/*/$i | wc -l) /4 ))"
done > read_counts_per_sample.txt 
```

We work with so many reads per sample:

Sample-DUMMY1_R1.fastq.gz: 44846\
Sample-DUMMY1_R2.fastq.gz: 44846\
Sample-DUMMY2_R1.fastq.gz: 19733\
Sample-DUMMY2_R2.fastq.gz: 19733\
Sample-DUMMY3_R1.fastq.gz: 19784\
Sample-DUMMY3_R2.fastq.gz: 19784\
Sample-DUMMY4_R1.fastq.gz: 24380\
Sample-DUMMY4_R2.fastq.gz: 24380

### Get read statistics

#### Run FastQC

```{bash}
mkdir -p results/fastqc 

#Submitted batch job 8100
sbatch scripts/run_fastqc.sh 

ll results/fastqc/*
```

Comments on the visual inspection of the HTML reports generated by FastQC:

-   The sequence quality looks good, most reads have a phred score between 28-40 when looking at the per base quality
-   The reads are typically 250 bp long (as expected)
-   In the over-represented sequences no hits to Illumina adapters where found

#### Run seqkit

```{bash}
mkdir -p results/seqkit 

#Submitted batch job 8114
sbatch scripts/seqkit.sh
```

Comments on the content:

-   Sample Sample-DUMMY1_R2.fastq.gz had reads of only 35 length, check with the sequencing center of something went wrong or if they did some quality cleaning as before sending the sequences
-   Other parameters look good

....

## Citation

If your work gets published and you provide the code for example on Github its useful to add information about how to cite your work

## Contact

For questions, please contact nd\@sds.net